test/utility.cpp

```diff
@@ -91,6 +91,129 @@ TEST_CASE("exchange", "[utility.exchange]") {
   }
 }
 
+// [forward], forward/move
+
+// template<class T>
+//   constexpr T&& forward(remove_reference_t<T>& t) noexcept;
+// template<class T>
+//   constexpr T&& forward(remove_reference_t<T>&& t) noexcept;
+TEST_CASE("forward", "[forward]") {
+  struct S {};
+
+  struct {
+    constexpr S operator()() const noexcept { return S{}; }
+  } make_s;
+
+  struct {
+    constexpr S const operator()() const noexcept { return S{}; }
+  } make_cs;
+
+  S s;
+  S const cs{};
+
+  // lvalue -> lvalue
+  CHECK(std::is_same<decltype(slb::forward<S&>(s)), S&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&>(s)), S const&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&>(cs)), S const&>::value);
+  CHECK(noexcept(slb::forward<S&>(s)));
+  CHECK(noexcept(slb::forward<S const&>(s)));
+  CHECK(noexcept(slb::forward<S const&>(cs)));
+
+  // lvalue -> rvalue
+  CHECK(std::is_same<decltype(slb::forward<S>(s)), S&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S&&>(s)), S&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const>(s)), S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&&>(s)), S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const>(cs)), S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&&>(cs)), S const&&>::value);
+  CHECK(noexcept(slb::forward<S>(s)));
+  CHECK(noexcept(slb::forward<S&&>(s)));
+  CHECK(noexcept(slb::forward<S const>(s)));
+  CHECK(noexcept(slb::forward<S const&&>(s)));
+  CHECK(noexcept(slb::forward<S const>(cs)));
+  CHECK(noexcept(slb::forward<S const&&>(cs)));
+
+  // rvalue -> rvalue
+  CHECK(std::is_same<decltype(slb::forward<S>(make_s())), S&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S&&>(make_s())), S&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const>(make_s())),
+                     S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&&>(make_s())),
+                     S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const>(make_cs())),
+                     S const&&>::value);
+  CHECK(std::is_same<decltype(slb::forward<S const&&>(make_cs())),
+                     S const&&>::value);
+  CHECK(noexcept(slb::forward<S>(make_s())));
+  CHECK(noexcept(slb::forward<S&&>(make_s())));
+  CHECK(noexcept(slb::forward<S const>(make_s())));
+  CHECK(noexcept(slb::forward<S const&&>(make_s())));
+  CHECK(noexcept(slb::forward<S const>(make_cs())));
+  CHECK(noexcept(slb::forward<S const&&>(make_cs())));
+
+  constexpr int x = slb::forward<int>(42);
+  CHECK(x == 42);
```

Comment:
Why is there a value check for rvalue -> rvalue but not for the other cases?

Reply:
Added more C++11 `constexpr` tests.